### PR TITLE
476: Fix javascript registration

### PR DESCRIPTION
**Pull request by toolstack**
Alternate solution to #476 which uses standard WP style and script registration and enqueuing functions.

### CODE REVIEW

File: gp-includes/assets-loader.php [hunk 1]
```diff
@@ -1,84 +1,94 @@
...
-	$scripts->add( 'gp-glossary', $url . '/glossary.js', array( 'gp-common' ), '20160329' );
-	$scripts->add( 'gp-translations-page', $url . '/translations-page.js', array( 'gp-common' ), '20150430' );
-	$scripts->add( 'mass-create-sets-page', $url . '/mass-create-sets-page.js', array( 'gp-common' ), '20150430' );
+	// Register our standard scripts.
+	wp_register_script( 'tablesorter', $url . '/jquery.tablesorter.min.js', array( 'jquery' ), '1.10.4' );
+	wp_register_script( 'gp-common', $url . '/common.js', array( 'jquery' ), '20150430' );
+	wp_register_script( 'gp-editor', $url . '/editor.js', array( 'gp-common', 'jquery-ui-tooltip' ), '20160329' );
+	wp_register_script( 'gp-glossary', $url . '/glossary.js', array( 'gp-common', 'gp-editor' ), '20160329' );
+	wp_register_script( 'gp-translations-page', $url . '/translations-page.js', array( 'gp-common', 'gp-editor' ), '20150430' );
+	wp_register_script( 'gp-mass-create-sets-page', $url . '/mass-create-sets-page.js', array( 'gp-common', 'gp-editor' ), '20150430' );
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: `gp-editor` has already `gp-common` as a dependency, so we don't have to repeat it for the others.

File: gp-templates/locales.php [hunk 2]
```diff
@@ -1,8 +1,7 @@
 <?php
 gp_title( __( 'Locales &lt; GlotPress', 'glotpress' ) );
 
-gp_enqueue_script( 'gp-common' );
-gp_enqueue_script( 'tablesorter' );
+gp_enqueue_script( array( 'gp-common', 'gp-editor', 'tablesorter' ) );
```
Comment (ocean90) [inline gp-templates/locales.php]: Why does it need `gp-editor`?

File: gp-templates/translations.php [hunk 3]
```diff
@@ -4,9 +4,7 @@
 	gp_project_links_from_root( $project ),
 	gp_link_get( $url, $translation_set->name ),
 ) );
-gp_enqueue_script( 'jquery-ui-core' );
-gp_enqueue_script( 'gp-editor' );
-gp_enqueue_script( 'gp-translations-page' );
+gp_enqueue_script( array( 'jquery-ui-core', 'gp-common', 'gp-editor', 'gp-translations-page' ) );
```
Comment (ocean90) [inline gp-templates/translations.php]: Which script needs `jquery-ui-core`? Can we add it as a dependency to the script?

File: gp-includes/assets-loader.php (hunk 1 above)
Comment (toolstack) [inline gp-includes/assets-loader.php]: I tossed back and forth a few times on how best to approach that, in the end I decided it was better to be verbose in the dependencies instead of nesting them.  Makes it clear what each script has for a dependency, however I'm fine with doing it the other way as well.

That's why in the template files I included all dependencies even though just including one would sometimes be all that was technically required.

File: gp-templates/locales.php (hunk 2 above)
Comment (toolstack) [inline gp-templates/locales.php]: It threw a JavaScript error on the page if it wasn't included, I didn't track it down any farther than that thought.

File: gp-templates/translations.php (hunk 3 above)
Comment (toolstack) [inline gp-templates/translations.php]: Not sure, it was on the page before and I didn't test to see if the dependencies was "real" or not :)

File: gp-includes/assets-loader.php (hunk 1 above)
Comment (ocean90) [inline gp-includes/assets-loader.php]: > in the end I decided it was better to be verbose in the dependencies instead of nesting them.

Well, if you want to go this route you have to define 'jquery' too. :)

Comment (toolstack) [inline gp-includes/assets-loader.php]: True, I'm not I'll update it to nest them.

File: gp-templates/locales.php (hunk 2 above)
Comment (toolstack) [inline gp-templates/locales.php]: Error doesn't occur any more without 'gp-editor'.  I'll yank the dependency.

File: gp-templates/translations.php (hunk 3 above)
Comment (toolstack) [inline gp-templates/translations.php]: Looks like it is not longer required, I'll remove it as well.

File: gp-includes/assets-loader.php [hunk 4]
```diff
@@ -1,84 +1,94 @@
...
+ *
+ * @package GlotPress
+ * @since 1.0.0
  */
 
 /**
- * Register the GlotPress styles
- *
- * @param WP_Styles $styles
+ * Register the GlotPress styles and load the base style sheet.
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: Docs are third-person singular elements, so third-person singular verbs should be used. (https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#language)

File: gp-includes/assets-loader.php [hunk 5]
```diff
@@ -1,84 +1,94 @@
...
  */
-
-function gp_enqueue_scripts() {
-	global $gp_enqueued_styles, $gp_enqueued_scripts;
-
-	if ( ! empty( $gp_enqueued_scripts ) ) {
-		foreach ( $gp_enqueued_scripts as $handle ) {
-			wp_enqueue_script( $handle );
-		}
+function gp_enqueue_style( $handles ) {
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: The function name should be in plural, `gp_enqueue_styles()`.

File: gp-includes/assets-loader.php [hunk 6]
```diff
@@ -1,84 +1,94 @@
...
+	// Check to see if $handles is an array, if not, then we can make it one to simplify the next loop.
+	if ( ! is_array( $handles ) ) {
+		$handles = array( $handles );
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: This can be simplified to  `$handles = (array) $handles;` or by adding `(array)` to the loop

`foreach ( (array) $handles as $handle ) {`

File: gp-includes/assets-loader.php [hunk 7]
```diff
@@ -1,84 +1,94 @@
...
 	}
 
-	if ( ! empty( $gp_enqueued_styles ) ) {
-		foreach ( $gp_enqueued_styles as $handle ) {
-			wp_enqueue_style( $handle );
-		}
+	// Loop through each handle we've been asked to enqueue.
+	foreach ( $handles as $handle ) {
+		// Store the handle name in the global array.
+		GP::$styles[] = $handle;
```
Comment (ocean90) [inline gp-includes/assets-loader.php:65]: Does this need a check whether `$handle` is already in the global array?

File: gp-includes/assets-loader.php [hunk 8]
```diff
@@ -1,84 +1,94 @@
...
-	global $gp_enqueued_styles;
-
-	$gp_enqueued_styles[] = $handle;
-}
+/**
+ * Enqueue one or more script.
+ *
+ * @param string|array $handles A single script handle to enqueue or an array or enqueue handles to enqueue.
+ */
+function gp_enqueue_script( $handles ) {
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: The function name should be in plural, `gp_enqueue_scripts()`.

File: gp-includes/assets-loader.php [hunk 9]
```diff
@@ -1,84 +1,94 @@
...
-
-function gp_enqueue_style( $handle ) {
-	global $gp_enqueued_styles;
-
-	$gp_enqueued_styles[] = $handle;
-}
+/**
+ * Enqueue one or more script.
+ *
+ * @param string|array $handles A single script handle to enqueue or an array or enqueue handles to enqueue.
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: Typo: The last 'or' should be 'of'.

File: gp-includes/assets-loader.php [hunk 10]
```diff
@@ -1,84 +1,94 @@
...
 
-add_action( 'wp_enqueue_scripts', 'gp_enqueue_scripts' );
-
-function gp_enqueue_style( $handle ) {
-	global $gp_enqueued_styles;
-
-	$gp_enqueued_styles[] = $handle;
-}
+/**
+ * Enqueue one or more script.
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: Typo: 'scripts', plural.

File: gp-includes/assets-loader.php [hunk 11]
```diff
@@ -1,84 +1,94 @@
...
+ *
+ * @param string|array $handles A single script handle to enqueue or an array or enqueue handles to enqueue.
+ */
+function gp_enqueue_script( $handles ) {
+	// Check to see if $handles is an array, if not, then we can make it one to simplify the next loop.
+	if ( ! is_array( $handles ) ) {
+		$handles = array( $handles );
+	}
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: See https://github.com/GlotPress/GlotPress-WP/pull/500/files#r77453961

File: gp-includes/assets-loader.php [hunk 12]
```diff
@@ -1,84 +1,94 @@
...
 }
 
+/**
+ * Print the styles that have been enqueued.
+ */
 function gp_print_styles() {
-	global $gp_enqueued_styles;
-	wp_print_styles( $gp_enqueued_styles );
+	// Only output the styles that GlotPress has registered, otherwise we'd be sending any style that the WordPress theme or plugins may have enqueued.
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: This comment should be in the DocBlock.

File: gp-includes/assets-loader.php [hunk 13]
```diff
@@ -1,84 +1,94 @@
...
+		GP::$scripts[] = $handle;
 
-	$gp_enqueued_scripts[] = $handle;
+		// Actually enqueue the handle via WordPress.
+		wp_enqueue_script( $handle );
+	}
 }
 
+/**
+ * Print the styles that have been enqueued.
```
Comment (ocean90) [inline gp-includes/assets-loader.php:107]: `@since` tag is missing.

File: gp-includes/assets-loader.php [hunk 14]
```diff
@@ -1,84 +1,94 @@
...
+	wp_print_styles( GP::$styles );
 }
 
+/**
+ * Print the scripts that have been enqueued.
+ */
 function gp_print_scripts() {
-	global $gp_enqueued_scripts;
-	wp_print_scripts( $gp_enqueued_scripts );
+	// Only output the scripts that GlotPress has registered, otherwise we'd be sending any scripts that the WordPress theme or plugins may have enqueued.
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: - This comment should be in the DocBlock.
- `@since` tag is missing.

File: gp-templates/header.php [hunk 15]
```diff
@@ -4,7 +4,11 @@
 	<meta http-equiv="Content-type" content="text/html; charset=utf-8" />
 	<title><?php echo gp_title(); ?></title>
 
-	<?php gp_head(); ?>
+	<?php
+	// Enqueue the base style so we don't have to load it manually on each page.
+	gp_enqueue_style( 'gp-base' );
```
Comment (ocean90) [inline gp-templates/header.php]: I wonder if this should be added via the `gp_head` action instead.

**Comment by ocean90** [comment]
What do you think about merging `gp_styles_default()` and `gp_register_scripts()` into `gp_register_assets()`?

File: gp-templates/header.php (hunk 15 above)
Comment (toolstack) [reply gp-templates/header.php]: Could, but leaving it in the header template makes it easier for sites to modify if required for their own custom templates.

File: gp-includes/assets-loader.php (hunk 8 above)
Comment (toolstack) [reply gp-includes/assets-loader.php]: I can change it, but it's going to break a lot of backwards compatibility with other plugins and custom templates.  If we change it (and the other one) should we include a wrapper to the singular version for backwards compatibility?

Comment (ocean90) [reply gp-includes/assets-loader.php]: OK, leave it singular.

**Comment by toolstack** [comment]
I did consider merging them but I think keeping them separate is probably better, though I don't have a really good reason for that.

**Comment by ocean90** [comment]
Note that `gp_register_scripts()`  was previously `gp_scripts_default()`, so you already changed one.

File: gp-includes/assets-loader.php [hunk 16]
```diff
@@ -1,84 +1,129 @@
...
+ *
+ * This function is a wrapper for backwards compatibility.
+ *
+ * @since 1.0.0
+ *
+ * @param string|array $handles A single script handle to enqueue or an array of enqueue handles to enqueue.
+ */
+function gp_enqueue_script( $handles ) {
+	gp_enqueue_scripts( $handles );
 }
```
Comment (ocean90) [inline gp-includes/assets-loader.php:104]: We shouldn't provide two functions which do the same. `gp_enqueue_script()` should continue to only support one handle.

``` php
function gp_enqueue_script( $handle ) {
    gp_enqueue_scripts( array( $handle ) );
}
```

Same for `gp_enqueue_style()`.

File: gp-includes/assets-loader.php [hunk 17]
```diff
@@ -60,7 +60,7 @@ function gp_enqueue_styles( $handles ) {
  * @param string $handle A single style handle to enqueue.
  */
 function gp_enqueue_style( $handle ) {
-	if ( ! in_array( $handle, GP::$styles ) ) {
+	if ( false === in_array( $handle, GP::$styles ) ) {
```
Comment (ocean90) [inline gp-includes/assets-loader.php]: What's the reason for this change? Did you mean to change it to`! in_array( $handle, GP::$styles, true )`?

Comment (toolstack) [reply gp-includes/assets-loader.php]: The coding standards check requires the comparison to be strict, hence the `false ===` instead of `!`.

Comment (ocean90) [reply gp-includes/assets-loader.php]: OK, that was for the third argument of `in_array()`. `in_array()` returns only boolean values so there is no need for the strict check.

Comment (toolstack) [reply gp-includes/assets-loader.php]: Yep, just noticed that, I've updated the PR but left in the `false ===`.

Comment (ocean90) [reply gp-includes/assets-loader.php]: Please remove it. :)

Comment (toolstack) [reply gp-includes/assets-loader.php]: Done.
